### DISCUSSION TITLE
SSR Dynamic Components

### DISCUSSION THREAD

**Discussion by lfre** [remix-run/remix#1502] [Q&A]
Currently, lazy components can only be achieved client-side due to the nature of `React.lazy` in React 17. This will change in React 18. Issue: https://github.com/remix-run/remix/issues/129

Example:

```js
// routes/index.jsx
import { Suspense, lazy } from "react";

const MyComponent = lazy(() => import('../components/lazy-component'));

export default function Index() {
  return (
    <div style={{ fontFamily: "system-ui, sans-serif", lineHeight: "1.4" }}>
      <h1>Welcome to Remix</h1>
      <Suspense>
         {/* This would be a ternary check to render conditionally */}
        <MyComponent />
      </Suspense>
      
// Omitted for brevity
```

Attempting to use `lazy` in a server-rendered component will throw the expected error:

```
Error: ReactDOMServer does not yet support Suspense.
```

Rendering different versions of components at the server is crucial for product teams to use with feature flags or A/B testing tools without requiring changing code. Being limited by doing this at the client-side takes away from Remix's biggest strengths.

Currently, the only workaround is to create multiple routes and change the URL that is sent to `RemixServer` without actually changing the URL displayed to the user. E.g:

With the routes:

/testing
  index.jsx
testing.a.jsx
testing.b.jsx

Rendering `/testing?version=a` will render internally `/testing/a` and `/testing/b` respectively, while maintaining the original `/testing` path.

It's a trivial example, but in practice, this would be using ENV variables or the response from an API.

```js
// entry.server.jsx

// Omitted for brevity

export default function handleRequest(
  request,
  responseStatusCode,
  responseHeaders,
  remixContext
) {
  const url = new URL(request.url);

  // This would be using process.env and/or querying a service to know which version to render
  const version = url.searchParams.has("version") && `/${url.searchParams.get('version')}`;

  if (version) {
    url.pathname = url.pathname + version;
  }
  
  const markup = renderToString(
    <RemixServer context={remixContext} url={url.href} />
  );
  
  // Omitted for brevity
}
```

While doable, this is by no means the desired solution.

Next provide this with `next/dynamic` through their own version of [react-loadable](https://github.com/jamiebuilds/react-loadable)

[Loadable Components](https://loadable-components.com/) is another possible solution.

What's your recommendation?

**Comment by sergiodxa** [comment]
I think if you use the React v18 it will work

**Reply by lfre** [reply]
@sergiodxa Yes, I presume it _should_ work, that's why I said: `This will change in React 18.`.

Running `npm install react@beta react-dom@beta` as [announced by the React team](https://github.com/reactwg/react-18/discussions/112) on a fresh `npx create-remix@latest` project fails the build with these errors when running `npm run dev`:

```bash
 > node_modules/@web-std/file/tsconfig.json:6:24: warning: Non-relative path "packages/file/src/lib.js" is not allowed when "baseUrl" is not set (did you forget a leading "./"?)
    6 │       "@web-std/file": ["packages/file/src/lib.js"]
      ╵                         ~~~~~~~~~~~~~~~~~~~~~~~~~~


 > node_modules/remix/esm/server.js:11:159: error: Could not resolve "@remix-run/server-runtime" (mark it as external to exclude it from the bundle)
    11 │ ...ookie, isSession, json, redirect } from '@remix-run/server-runtime';
       ╵                                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Regardless, the hope with React 18 is that this will __just work__ 🤞  so as someone who's considering switching a production application to Remix is important to know if the recommendation is to wait for React 18 stable vs expecting a solution within Remix in the near future.

Appreciate your response nonetheless!
